### ISSUE TITLE
FP Syntax Error: match guards w struct literals

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-analyzer#23050]
see also https://github.com/rust-lang/rust-analyzer/issues/15487

The following code
```rust
//@ check-pass

// Unlike `if` condition, `match` guards accept struct literals.
// This is detected in <https://github.com/rust-lang/rust/pull/74566#issuecomment-663613705>.

#[derive(PartialEq)]
struct Foo {
    x: isize,
}

fn foo(f: Foo) {
    match () {
        () if f == Foo { x: 42 } => {}
        () if let Foo { x: 0.. } = Foo { x: 42 } => {}
        _ => {}
    }
}

fn main() {}

```

comiled with rustc (cargo check):
```
    Checking _struct-literal-in-match-guard v0.1.0 (/tmp/icemaker_global_tempdir.TzJ5JP3WYP2D/rustc_testrunner_tmpdir.qb2aEcwfiCyt/_struct-literal-in-match-guard)
warning: struct `Foo` is never constructed
 --> src/main.rs:7:8
  |
7 | struct Foo {
  |        ^^^
  |
  = note: `#[warn(dead_code)]` (part of `#[warn(unused)]`) on by default

warning: function `foo` is never used
  --> src/main.rs:11:4
   |
11 | fn foo(f: Foo) {
   |    ^^^

warning: `_struct-literal-in-match-guard` (bin "_struct-literal-in-match-guard") generated 2 warnings
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.14s

```

but errors were issued by `rust-analyzer diagnostics`
```
at crate _struct_literal_in_match_guard, file /tmp/icemaker_global_tempdir.TzJ5JP3WYP2D/rustc_testrunner_tmpdir.qb2aEcwfiCyt/_struct-literal-in-match-guard/src/main.rs: Error SyntaxError from LineCol { line: 13, col: 38 } to LineCol { line: 13, col: 38 }: Syntax Error: expected FAT_ARROW
at crate _struct_literal_in_match_guard, file /tmp/icemaker_global_tempdir.TzJ5JP3WYP2D/rustc_testrunner_tmpdir.qb2aEcwfiCyt/_struct-literal-in-match-guard/src/main.rs: Error SyntaxError from LineCol { line: 13, col: 42 } to LineCol { line: 13, col: 42 }: Syntax Error: expected SEMICOLON
at crate _struct_literal_in_match_guard, file /tmp/icemaker_global_tempdir.TzJ5JP3WYP2D/rustc_testrunner_tmpdir.qb2aEcwfiCyt/_struct-literal-in-match-guard/src/main.rs: Error SyntaxError from LineCol { line: 13, col: 42 } to LineCol { line: 13, col: 42 }: Syntax Error: expected expression
at crate _struct_literal_in_match_guard, file /tmp/icemaker_global_tempdir.TzJ5JP3WYP2D/rustc_testrunner_tmpdir.qb2aEcwfiCyt/_struct-literal-in-match-guard/src/main.rs: Error SyntaxError from LineCol { line: 13, col: 49 } to LineCol { line: 13, col: 49 }: Syntax Error: expected pattern
at crate _struct_literal_in_match_guard, file /tmp/icemaker_global_tempdir.TzJ5JP3WYP2D/rustc_testrunner_tmpdir.qb2aEcwfiCyt/_struct-literal-in-match-guard/src/main.rs: Error RustcHardError("E0425") from LineCol { line: 13, col: 35 } to LineCol { line: 13, col: 38 }: no such value in this scope
at crate _struct_literal_in_match_guard, file /tmp/icemaker_global_tempdir.TzJ5JP3WYP2D/rustc_testrunner_tmpdir.qb2aEcwfiCyt/_struct-literal-in-match-guard/src/main.rs: Error RustcHardError("E0425") from LineCol { line: 13, col: 41 } to LineCol { line: 13, col: 42 }: no such value in this scope
at crate _struct_literal_in_match_guard, file /tmp/icemaker_global_tempdir.TzJ5JP3WYP2D/rustc_testrunner_tmpdir.qb2aEcwfiCyt/_struct-literal-in-match-guard/src/main.rs: Error RustcHardError("E0308") from LineCol { line: 13, col: 44 } to LineCol { line: 13, col: 46 }: expected (), found i32

diagnostic scan complete


Error: diagnostic error detected

Stack backtrace:
   0: <anyhow::Error>::msg::<&str>
             at /home/matthias/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/anyhow-1.0.102/src/backtrace.rs:10:14
   1: anyhow::__private::format_err
             at /home/matthias/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/anyhow-1.0.102/src/lib.rs:687:13
   2: <rust_analyzer::cli::flags::Diagnostics>::run_
             at /home/matthias/vcs/github/rust-analyzer/crates/rust-analyzer/src/cli/diagnostics.rs:123:13
   3: <rust_analyzer::cli::flags::Diagnostics>::run::{closure#0}
             at /home/matthias/vcs/github/rust-analyzer/crates/rust-analyzer/src/cli/diagnostics.rs:20:29
   4: <stdx::thread::Builder>::spawn::<<rust_analyzer::cli::flags::Diagnostics>::run::{closure#0}, core::result::Result<(), anyhow::Error>>::{closure#0}
             at /home/matthias/vcs/github/rust-analyzer/crates/stdx/src/thread.rs:74:13
   5: std::sys::backtrace::__rust_begin_short_backtrace::<<stdx::thread::Builder>::spawn<<rust_analyzer::cli::flags::Diagnostics>::run::{closure#0}, core::result::Result<(), anyhow::Error>>::{closure#0}, core::result::Result<(), anyhow::Error>>
             at /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/backtrace.rs:166:18
   6: std::thread::lifecycle::spawn_unchecked::<<stdx::thread::Builder>::spawn<<rust_analyzer::cli::flags::Diagnostics>::run::{closure#0}, core::result::Result<(), anyhow::Error>>::{closure#0}, core::result::Result<(), anyhow::Error>>::{closure#1}::{closure#0}
             at /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/lifecycle.rs:70:13
   7: <core::panic::unwind_safe::AssertUnwindSafe<std::thread::lifecycle::spawn_unchecked<<stdx::thread::Builder>::spawn<<rust_analyzer::cli::flags::Diagnostics>::run::{closure#0}, core::result::Result<(), anyhow::Error>>::{closure#0}, core::result::Result<(), anyhow::Error>>::{closure#1}::{closure#0}> as core::ops::function::FnOnce<()>>::call_once
             at /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panic/unwind_safe.rs:275:9
   8: std::panicking::catch_unwind::do_call::<core::panic::unwind_safe::AssertUnwindSafe<std::thread::lifecycle::spawn_unchecked<<stdx::thread::Builder>::spawn<<rust_analyzer::cli::flags::Diagnostics>::run::{closure#0}, core::result::Result<(), anyhow::Error>>::{closure#0}, core::result::Result<(), anyhow::Error>>::{closure#1}::{closure#0}>, core::result::Result<(), anyhow::Error>>
             at /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:576:43
   9: std::panicking::catch_unwind::<core::result::Result<(), anyhow::Error>, core::panic::unwind_safe::AssertUnwindSafe<std::thread::lifecycle::spawn_unchecked<<stdx::thread::Builder>::spawn<<rust_analyzer::cli::flags::Diagnostics>::run::{closure#0}, core::result::Result<(), anyhow::Error>>::{closure#0}, core::result::Result<(), anyhow::Error>>::{closure#1}::{closure#0}>>
             at /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:544:19
  10: std::panic::catch_unwind::<core::panic::unwind_safe::AssertUnwindSafe<std::thread::lifecycle::spawn_unchecked<<stdx::thread::Builder>::spawn<<rust_analyzer::cli::flags::Diagnostics>::run::{closure#0}, core::result::Result<(), anyhow::Error>>::{closure#0}, core::result::Result<(), anyhow::Error>>::{closure#1}::{closure#0}>, core::result::Result<(), anyhow::Error>>
             at /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:359:14
  11: std::thread::lifecycle::spawn_unchecked::<<stdx::thread::Builder>::spawn<<rust_analyzer::cli::flags::Diagnostics>::run::{closure#0}, core::result::Result<(), anyhow::Error>>::{closure#0}, core::result::Result<(), anyhow::Error>>::{closure#1}
             at /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/lifecycle.rs:68:26
  12: <std::thread::lifecycle::spawn_unchecked<<stdx::thread::Builder>::spawn<<rust_analyzer::cli::flags::Diagnostics>::run::{closure#0}, core::result::Result<(), anyhow::Error>>::{closure#0}, core::result::Result<(), anyhow::Error>>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
             at /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5
  13: <alloc::boxed::Box<dyn core::ops::function::FnOnce<(), Output = ()> + core::marker::Send> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/1ed2df61a19042f231709eb05d032ae9e2cb2084/library/alloc/src/boxed.rs:2318:9
  14: <std::sys::thread::unix::Thread>::new::thread_start
             at /rustc/1ed2df61a19042f231709eb05d032ae9e2cb2084/library/std/src/sys/thread/unix.rs:123:17
  15: <unknown>
  16: <unknown>

```

version information
```
rust-analyzer 0.0.0 (6817dd3094 2026-08-05)

```
